app/uk/gov/hmrc/mtdsatestsupportapi/controllers/requestParsers/validators/CreateTestBusinessValidator.scala

```diff
@@ -90,32 +90,88 @@ class CreateTestBusinessValidator @Inject() (clock: Clock, featureSwitches: Feat
       validateMissingPostcode(business) ++
         validateAccountingPeriod(business) ++
+        validateBusinessAddress(business) ++
+        validateTradingName(business) ++
         business.commencementDate.map(validateCommencementDate).getOrElse(Nil)
 
     errorsResult(errors)
   }
 
-  private def validateMissingPostcode(business: Business): List[MtdError] = {
-    (business.businessAddressPostcode, business.businessAddressCountryCode) match {
-      case (None, Some("GB")) => List(MissingPostcodeError)
-      case _                  => NoValidationErrors
+  private def validateMissingPostcode(business: Business): Seq[MtdError] = {
+    if (featureSwitches.isEnabled("release5")) {
+      if (requiresBusinessPostcode(business) && business.businessAddressPostcode.isEmpty) {
+        Seq(MissingPostcodeError)
+      } else {
+        Nil
+      }
+    } else {
+      (business.businessAddressPostcode, business.businessAddressCountryCode) match {
+        case (None, Some("GB")) => List(MissingPostcodeError)
+        case _                  => NoValidationErrors
+      }
     }
   }
 
-  private def validateAccountingPeriod(business: Business): Seq[MtdError] =
-    if (featureSwitches.isEnabled("release5"))
+  private def validateAccountingPeriod(business: Business): Seq[MtdError] = {
+    if (featureSwitches.isEnabled("release5")) {
       (business.firstAccountingPeriodStartDate, business.firstAccountingPeriodEndDate) match {
         case (Some(start), Some(end)) => TaxYearAlignmentDateRangeValidation.validate(start, end, RuleFirstAccountingDateRangeInvalid)
         case (None, Some(_))          => Seq(MissingFirstAccountintPeriodStartDateError)
         case (Some(_), None)          => Seq(MissingFirstAccountintPeriodEndDateError)
         case _                        => Nil
       }
-    else Nil
+    } else {
+      Nil
+    }
+  }
 
   private def validateCommencementDate(date: LocalDate): List[MtdError] = {
     val today = LocalDate.now(clock)
 
     if (date < today) Nil else List(RuleCommencementDateNotSupported)
   }
 
+  private def validateTradingName(business: Business): Seq[MtdError] = {
+    if (featureSwitches.isEnabled("release5")) {
+      if (forbidsTradingName(business) && business.tradingName.nonEmpty) {
+        Seq(RuleUnexpectedTradingName)
+      } else if (requiresTradingName(business) && business.tradingName.isEmpty) {
+        Seq(RuleMissingTradingName)
+      } else {
+        Nil
+      }
+    } else {
+      Nil
+    }
+  }
+
+  private def validateBusinessAddress(business: Business): Seq[MtdError] = {
+    def haveSufficientBusinessAddress = business.businessAddressLineOne.nonEmpty && business.businessAddressCountryCode.nonEmpty
+
+    if (featureSwitches.isEnabled("release5")) {
+      if (forbidsBusinessAddress(business) && business.hasAnyBusinessAddressDetails) {
+        Seq(RuleUnexpectedBusinessAddress)
+      } else if (requiresBusinessAddress(business) && !haveSufficientBusinessAddress) {
```

Comment:
Happy for you to leave as-is (as may be a personal preference thing) but as `haveSufficientBusinessAddress` is only used once here and is being used for its opposite purpose, would just consider inlining it for its intended purpose, e.g.: `else if (requiresBusinessAddress(business) && (business.businessAddressLineOne.isEmpty || business.businessAddressCountryCode.isEmpty)`. Or alternatively rename `haveSufficientBusinessAddress` to `hasInsufficientBusinessAddress` (or wtte).

Reply:
I personally find it harder to follow pre-negated methods, usually. Especially when it comes to validation. I end up reading the `else` blocks as double negations etc.

Happy to change it if you insist. We might be reading the code differently. I tried to write it so that each bit makes sense in isolation, without having to go to the method. Maybe I failed.

Reply:
I am sure this is just a preference thing, but thought i'd give my opinion to see what you thought. Happy for you to leave it if you find it better 👍